
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Drug-Target Interaction - TDC</title>
    <link rel="stylesheet" href="/TDC/assets/css/app.css">
    <link rel="shortcut icon" type="image/png"
           href="/TDC/favicon.png" 
    />
    <script defer src="https://use.fontawesome.com/releases/v5.3.1/js/all.js"></script>
    <!-- Begin Jekyll SEO tag v2.6.1 -->
<title>Drug-Target Interaction | TDC</title>
<meta name="generator" content="Jekyll v3.8.6" />
<meta property="og:title" content="Drug-Target Interaction" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Datasets for Medicinal Machine Learning" />
<meta property="og:description" content="Datasets for Medicinal Machine Learning" />
<link rel="canonical" href="http://localhost:4000/TDC/multi_pred_tasks/dti/" />
<meta property="og:url" content="http://localhost:4000/TDC/multi_pred_tasks/dti/" />
<meta property="og:site_name" content="TDC" />
<script type="application/ld+json">
{"url":"http://localhost:4000/TDC/multi_pred_tasks/dti/","headline":"Drug-Target Interaction","description":"Datasets for Medicinal Machine Learning","@type":"WebPage","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-183522810-1"></script>
<script>
  window['ga-disable-UA-183522810-1'] = window.doNotTrack === "1" || navigator.doNotTrack === "1" || navigator.doNotTrack === "yes" || navigator.msDoNotTrack === "1";
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'UA-183522810-1');
</script><!-- head scripts --></head>

  <body>
    
<nav class="navbar is-primary" >
    <div class="container">
        <div class="navbar-brand">
            <a class="navbar-brand" href="/TDC/">
            <span><img src="/TDC/logonav.png" alt="Logo" style="height: auto; width: auto; max-height: 45px; max-width: 250px;"></span>
            </a>
            <a role="button" class="navbar-burger burger" aria-label="menu" aria-expanded="false" data-target="navMenu">
                <span aria-hidden="true"></span>
                <span aria-hidden="true"></span>
                <span aria-hidden="true"></span>
            </a>
        </div>
        <div class="navbar-menu" id="navMenu">
            <div class="navbar-end">
                
                
                    
                    <a href="/TDC/" class="navbar-item ">Home</a>
                    
                
                    
                    <a href="/TDC/start/" class="navbar-item ">Quick Start</a>
                    
                
                    
                    <a href="/TDC/news/" class="navbar-item ">News</a>
                    
                
                    
                    <div class="navbar-item has-dropdown is-hoverable">
                        <a href="/TDC/overview/" class="navbar-link ">Datasets</a>
                        <div class="navbar-dropdown">
                            
                            <a href="/TDC/overview/" class="navbar-item ">Overview</a>
                            
                            <a href="/TDC/single_pred_tasks/overview" class="navbar-item ">Single-instance Prediction</a>
                            
                            <a href="/TDC/multi_pred_tasks/overview" class="navbar-item ">Multi-instance Prediction</a>
                            
                            <a href="/TDC/generation_tasks/overview" class="navbar-item ">Generation</a>
                            
                        </div>
                    </div>
                    
                
                    
                    <div class="navbar-item has-dropdown is-hoverable">
                        <a href="/TDC/fct_overview/" class="navbar-link ">Data Functions</a>
                        <div class="navbar-dropdown">
                            
                            <a href="/TDC/fct_overview/" class="navbar-item ">Overview</a>
                            
                            <a href="/TDC/functions/data_evaluation/" class="navbar-item ">Model Evaluation</a>
                            
                            <a href="/TDC/functions/data_process/" class="navbar-item ">Data Processing</a>
                            
                            <a href="/TDC/functions/data_split/" class="navbar-item ">Data Split</a>
                            
                            <a href="/TDC/functions/oracles/" class="navbar-item ">Molecule Generation Oracles</a>
                            
                        </div>
                    </div>
                    
                
                    
                    <a href="/TDC/team/" class="navbar-item ">Team</a>
                    
                
                <a href="https://github.com/mims-harvard/TDC" class="navbar-item">GitHub</a>
                
            </div>

        </div>
    </div>
</nav>

    
    


    <section class="section">
        <div class="container">
            <div class="columns">
                
                <div class="column is-4-desktop is-4-tablet">
                    

<aside class="menu">

    <p class="menu-label"></p>
    <ul class="menu-list">
        
        <li>
            <a href="/TDC/multi_pred_tasks/overview/" class=""><strong>Multi-instance Prediction Problem</strong></a>
            
            <ul>
                
                
                <li><a href="/TDC/multi_pred_tasks/dti/" class="is-active"><code>DTI</code> Drug-Target Interaction</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/ddi/" class=""><code>DDI</code> Drug-Drug Interaction</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/ppi/" class=""><code>PPI</code> Protein-Protein Interaction</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/gdi/" class=""><code>GDA</code> Gene-Disease Association</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/drugres/" class=""><code>DrugRes</code> Drug Response Prediction</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/drugsyn/" class=""><code>DrugSyn</code> Drug Synergy Prediction</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/peptidemhc/" class=""><code>PeptideMHC</code> Peptide-MHC Binding</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/antibodyaff/" class=""><code>AntibodyAff</code> Antibody-Antigen Affinity</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/mti/" class=""><code>MTI</code> miRNA-Target Interaction</a></li>
                
                
                
                <li><a href="/TDC/multi_pred_tasks/catalyst/" class=""><code>Catalyst</code> Catalyst Prediction</a></li>
                
                
            </ul>
            
        </li>
            
    </ul>

    <p class="menu-label"></p>
    <ul class="menu-list">
        
        <li>
            <a href="/TDC/overview/" class=""><strong>Others</strong></a>
            
            <ul>
                
                
                <li><a href="/TDC/single_pred_tasks/overview/" class="">Single-instance Prediction Problem</a></li>
                
                
                
                <li><a href="/TDC/generation_tasks/overview/" class="">Generation Problem</a></li>
                
                
            </ul>
            
        </li>
            
    </ul>

</aside>
                </div>
                
                <div class="column is-8">
                    
                    
                    
                    
    
    

<div class="contents">
    <div class="menu">
        <p class="menu-label">Dataset Index</p>
        <ul class="menu-list">
  <li><a href="#bindingdb">BindingDB</a></li>
  <li><a href="#davis">DAVIS</a></li>
  <li><a href="#kiba">KIBA</a></li>
</ul>
    </div>
</div>




<div class="content">
    <h3 id="bindingdb">BindingDB</h3>

<p class="is-size-6">  <strong> Dataset Description: </strong> BindingDB is a public, web-accessible database of measured binding affinities, focusing chiefly on the interactions of protein considered to be drug-targets with small, drug-like molecules. </p>

<p class="is-size-6">  <strong> Task Description: </strong>Regression. Given the target amino acid sequence/compound SMILES string, predict their binding affinity. </p>

<p class="is-size-6">  <strong> Dataset Statistics: </strong> (# of DTI pairs, # of drugs, # of proteins) 52,284/10,665/1,413 for Kd, 991,486/549,205/5,078 for IC50, 375,032/174,662/3,070 for Ki. </p>

<p class="is-size-6">  <strong> Dataset Split: </strong> <span class="tag is-info is-light">Random Split</span> <span class="tag is-info is-light">Cold Drug Split</span> <span class="tag is-info is-light">Cold Protein Split</span> </p>

<p class="is-size-6">  <strong> Note: </strong> BindingDB is the collection of many assays. Since different assays use different metrics, TDC separates them as separate datasets. Specifically, it has four datasets with Kd, IC50, Ki as the units. </p>

<p class="is-size-6"> <strong>Tips:</strong> Checkout the <a href="/TDC/functions/data_process">data processing</a> page for binarization, label distribution visualization, edge list/DGL/PyTorch graph transformation. </p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">tdc.multi_pred</span> <span class="kn">import</span> <span class="n">DTI</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">DTI</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s">'BindingDB_Kd'</span><span class="p">)</span>
<span class="c1"># data = DTI(name = 'BindingDB_IC50')
# data = DTI(name = 'BindingDB_Ki')
</span><span class="n">split</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_split</span><span class="p">()</span>
</code></pre></div></div>

<p class="is-size-6">  <strong> References: </strong>  </p>

<p><a href="https://academic.oup.com/nar/article-abstract/35/suppl_1/D198/1119109">[1] Liu, Tiqing, et al. “BindingDB: a web-accessible database of experimentally determined protein–ligand binding affinities.” Nucleic acids research 35.suppl_1 (2007): D198-D201. </a></p>

<p><a href="https://arxiv.org/abs/2004.08919">[2] Huang, Kexin, et al. “DeepPurpose: a Deep Learning Library for Drug-Target Interaction Prediction” Bioinformatics. </a></p>

<p class="is-size-6"> <strong> Dataset License: </strong> <a href="https://creativecommons.org/licenses/by/3.0/us/">CC BY 3.0 US</a>.</p>

<hr />

<h3 id="davis">DAVIS</h3>

<p class="is-size-6">  <strong> Dataset Description: </strong> The interaction of 72 kinase inhibitors with 442 kinases covering &gt;80% of the human catalytic protein kinome. </p>

<p class="is-size-6">  <strong> Task Description: </strong>Regression. Given the target amino acid sequence/compound SMILES string, predict their binding affinity. </p>

<p class="is-size-6">  <strong> Dataset Statistics: </strong> 27,621 DTI pairs, 68 drugs, 379 proteins. </p>

<p class="is-size-6">  <strong> Dataset Split: </strong> <span class="tag is-info is-light">Random Split</span> <span class="tag is-info is-light">Cold Drug Split</span> <span class="tag is-info is-light">Cold Protein Split</span> </p>

<p class="is-size-6"> <strong>Tips:</strong> Checkout the <a href="/TDC/functions/data_process">data processing</a> page for binarization, label distribution visualization, edge list/DGL/PyTorch graph transformation. </p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">tdc.multi_pred</span> <span class="kn">import</span> <span class="n">DTI</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">DTI</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s">'DAVIS'</span><span class="p">)</span>
<span class="n">split</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_split</span><span class="p">()</span>
</code></pre></div></div>

<p class="is-size-6">  <strong> References: </strong>  </p>

<p><a href="https://www.nature.com/articles/nbt.1990">[1] Davis, M., Hunt, J., Herrgard, S. et al. Comprehensive analysis of kinase inhibitor selectivity. Nat Biotechnol 29, 1046–1051 (2011). </a></p>

<p><a href="https://arxiv.org/abs/2004.08919">[2] Huang, Kexin, et al. “DeepPurpose: a Deep Learning Library for Drug-Target Interaction Prediction” Bioinformatics. </a></p>

<p class="is-size-6"> <strong> Dataset License: </strong> Not Specified. <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>.</p>

<hr />

<h3 id="kiba">KIBA</h3>

<p class="is-size-6">  <strong> Dataset Description: </strong> Toward making use of the complementary information captured by the various bioactivity types, including IC50, K(i), and K(d), Tang et al. introduces a model-based integration approach, termed KIBA to generate an integrated drug-target bioactivity matrix. </p>

<p class="is-size-6">  <strong> Task Description: </strong>Regression. Given the target amino acid sequence/compound SMILES string, predict their binding affinity. </p>

<p class="is-size-6">  <strong> Dataset Statistics: </strong> 118,036 DTI pairs, 2,068 drugs, 229 proteins. </p>

<p class="is-size-6">  <strong> Dataset Split: </strong> <span class="tag is-info is-light">Random Split</span> <span class="tag is-info is-light">Cold Drug Split</span> <span class="tag is-info is-light">Cold Protein Split</span> </p>

<p class="is-size-6"> <strong>Tips:</strong> Checkout the <a href="/TDC/functions/data_process">data processing</a> page for binarization, label distribution visualization, edge list/DGL/PyTorch graph transformation. </p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">tdc.multi_pred</span> <span class="kn">import</span> <span class="n">DTI</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">DTI</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s">'KIBA'</span><span class="p">)</span>
<span class="n">split</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_split</span><span class="p">()</span>
</code></pre></div></div>

<p class="is-size-6">  <strong> References: </strong>  </p>

<p><a href="https://pubmed.ncbi.nlm.nih.gov/24521231/">[1] Tang J, Szwajda A, Shakyawar S, et al. Making sense of large-scale kinase inhibitor bioactivity data sets: a comparative and integrative analysis. J Chem Inf Model. 2014;54(3):735-743. </a></p>

<p><a href="https://arxiv.org/abs/2004.08919">[2] Huang, Kexin, et al. “DeepPurpose: a Deep Learning Library for Drug-Target Interaction Prediction” Bioinformatics. </a></p>

<p class="is-size-6"> <strong> Dataset License: </strong> Not Specified. <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>.</p>

<hr />


</div>
                </div>
                
            </div>
        </div>
    </section>
    
        <footer class="footer">
    <div class="container">
        
        <div class="columns is-mobile">
            <div class="column is-8 has-text-left is-vcentered">
                <a class="navbar-brand" href="/TDC/">
                    <span><img src="/TDC/tdc_horizontal.png" alt="Logo" style="max-height: 40px; max-width: 250px;"></span>
                </a>
            </div>
            <div class="column is-4 has-text-right is-vcentered">
                <a href="https://github.com/mims-harvard/TDC">
                    <span class="icon is-large">
                      <i class="fas fab fa-github fa-3x"></i>
                    </span>
                </a>
            </div>
        </div>
        
    </div>
</footer>
    
    <script src="/TDC/assets/js/app.js" type="text/javascript"></script><!-- footer scripts --></body>
</html>

