
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Protein-Peptide Binding Affinity Benchmark Group - TDC</title>
    <link rel="stylesheet" href="/assets/css/app.css">
    <link rel="shortcut icon" type="image/png"
           href="/favicon.png" 
    />
    <script defer src="https://use.fontawesome.com/releases/v5.3.1/js/all.js"></script>
    <!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Protein-Peptide Binding Affinity Benchmark Group | TDC</title>
<meta name="generator" content="Jekyll v3.9.3" />
<meta property="og:title" content="Protein-Peptide Binding Affinity Benchmark Group" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Artificial intelligence foundation for therapeutic science" />
<meta property="og:description" content="Artificial intelligence foundation for therapeutic science" />
<link rel="canonical" href="http://localhost:4000/benchmark/proteinpeptide_group/overview/" />
<meta property="og:url" content="http://localhost:4000/benchmark/proteinpeptide_group/overview/" />
<meta property="og:site_name" content="TDC" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Protein-Peptide Binding Affinity Benchmark Group" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebPage","description":"Artificial intelligence foundation for therapeutic science","headline":"Protein-Peptide Binding Affinity Benchmark Group","url":"http://localhost:4000/benchmark/proteinpeptide_group/overview/"}</script>
<!-- End Jekyll SEO tag -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-183522810-1"></script>
<script>
  window['ga-disable-UA-183522810-1'] = window.doNotTrack === "1" || navigator.doNotTrack === "1" || navigator.doNotTrack === "yes" || navigator.msDoNotTrack === "1";
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'UA-183522810-1');
</script><!-- head scripts --></head>

  <body>
    
<nav class="navbar is-primary" >
    <div class="container">
        <div class="navbar-brand">
            <a class="navbar-brand" href="/">
            <span><img src="/logonav.png" alt="Logo" style="height: auto; width: auto; max-height: 45px; max-width: 250px;"></span>
            </a>
            <a role="button" class="navbar-burger burger" aria-label="menu" aria-expanded="false" data-target="navMenu">
                <span aria-hidden="true"></span>
                <span aria-hidden="true"></span>
                <span aria-hidden="true"></span>
            </a>
        </div>
        <div class="navbar-menu" id="navMenu">
            <div class="navbar-end">
                
                
                    
                    <a href="/" class="navbar-item ">Home</a>
                    
                
                    
                    <a href="/start/" class="navbar-item ">Start</a>
                    
                
                    
                    <div class="navbar-item has-dropdown is-hoverable">
                        <a href="/overview/" class="navbar-link ">Datasets</a>
                        <div class="navbar-dropdown">
                            
                            <a href="/overview/" class="navbar-item ">Overview</a>
                            
                            <a href="/single_pred_tasks/overview" class="navbar-item ">Single-instance Prediction</a>
                            
                            <a href="/multi_pred_tasks/overview" class="navbar-item ">Multi-instance Prediction</a>
                            
                            <a href="/generation_tasks/overview" class="navbar-item ">Generation</a>
                            
                        </div>
                    </div>
                    
                
                    
                    <div class="navbar-item has-dropdown is-hoverable">
                        <a href="/fct_overview/" class="navbar-link ">Data Functions</a>
                        <div class="navbar-dropdown">
                            
                            <a href="/fct_overview/" class="navbar-item ">Overview</a>
                            
                            <a href="/functions/data_evaluation/" class="navbar-item ">Model Evaluation</a>
                            
                            <a href="/functions/data_split/" class="navbar-item ">Dataset Splits</a>
                            
                            <a href="/functions/data_process/" class="navbar-item ">Data Processing</a>
                            
                            <a href="/functions/oracles/" class="navbar-item ">Molecule Generation Oracles</a>
                            
                        </div>
                    </div>
                    
                
                    
                    <div class="navbar-item has-dropdown is-hoverable">
                        <a href="/benchmark/overview/" class="navbar-link ">Leaderboards</a>
                        <div class="navbar-dropdown">
                            
                            <a href="/benchmark/overview/" class="navbar-item ">Guidelines</a>
                            
                            <a href="/benchmark/admet_group/overview" class="navbar-item ">ADMET Group</a>
                            
                            <a href="/benchmark/drugcombo_group/overview" class="navbar-item ">DrugCombo Group</a>
                            
                            <a href="/benchmark/docking_group/overview" class="navbar-item ">Docking Group</a>
                            
                            <a href="/benchmark/dti_dg_group/overview" class="navbar-item ">DTI DG Group</a>
                            
                            <a href="/benchmark/scdti_group/overview" class="navbar-item ">Single-cell DTI Group</a>
                            
                            <a href="/benchmark/proteinpeptide_group/overview" class="navbar-item ">Protein-Peptide Binding Affinity Group</a>
                            
                            <a href="/benchmark/counterfactual_group/overview" class="navbar-item ">Counterfactual Prediction Group</a>
                            
                        </div>
                    </div>
                    
                
                    
                    <a href="https://huggingface.co/tdc" class="navbar-item ">HF Models</a>
                    
                
                    
                    <a href="/news/" class="navbar-item ">News</a>
                    
                
                    
                    <a href="/team/" class="navbar-item ">Team</a>
                    
                
                <a href="https://arxiv.org/abs/2102.09548" class="navbar-item">NeurIPS Paper</a>
                <a href="https://www.nature.com/articles/s41589-022-01131-2" class="navbar-item">Nat Chem Bio Paper</a>
                <a href="https://tdc.readthedocs.io" class="navbar-item">Docs</a>
                <a href="https://github.com/mims-harvard/TDC" class="navbar-item">GitHub</a>
                
            </div>

        </div>
    </div>
</nav>

    
    


    <section class="section">
        <div class="container">
            <div class="columns">
                
                <div class="column is-4-desktop is-4-tablet">
                    

<aside class="menu">

    <p class="menu-label"></p>
    <ul class="menu-list">
        
        <li>
            <a href="/benchmark/proteinpeptide_group/overview" class=""><strong>Protein-Peptide Binding Affinity Group Leaderboard</strong></a>
            
            <ul>
                
                
                <li><a href="/benchmark/proteinpeptide_group/brown_mdm2_ace2_12ca5/" class=""><code>TDC.brown_mdm2_ace2_12ca5</code> Leaderboard</a></li>
                
                
            </ul>
            
        </li>
            
    </ul>

    <p class="menu-label"></p>
    <ul class="menu-list">
        
        <li>
            <a href="" class=""><strong>References</strong></a>
            
            <ul>
                
                
                <li><a href="/benchmark/overview/#step-by-step-instructions" class="">Submission Instructions</a></li>
                
                
                
                <li><a href="/benchmark/overview/#further-details-about-benchmark-groups-in-tdc" class="">BenchmarkGroup Class Documentation</a></li>
                
                
                
                <li><a href="/benchmark/overview/" class="">Leaderboard Guidelines</a></li>
                
                
                
                <li><a href="https://dataverse.harvard.edu/file.xhtml?fileId=9649623&version=85.0" class="">Dataset Description</a></li>
                
                
            </ul>
            
        </li>
            
    </ul>

</aside>
                </div>
                
                <div class="column is-8">
                    
                    
                    
                    
<div class="content">
    <h2 id="protein-peptide-binding-affinity-benchmark-group">Protein-Peptide Binding Affinity Benchmark Group</h2>

<p class="is-size-5">Despite the availability of several benchmarks for protein-protein interactions, there is a comparatively269
large gap compared to benchmarks for protein-peptide binding affinity prediction. For example, the270
renowned multi-task benchmark for Protein sEquence undERstanding (PEER) [ 24], lacks support271
for a specifically defined protein-peptide binding affinity prediction task. Protein-peptide binding272
affinity prediction and protein-protein binding affinity prediction involve similar underlying biological273
interactions, but they differ significantly in their complexity and the methods used to predict them [ 2].274
Here’s why they are treated differently:275
• Molecular Complexity and Size: Proteins are generally large and complex molecules276
composed of multiple domains and potentially numerous subunits. Peptides, however, are277
smaller and typically consist of only a short chain of amino acids. The size difference affects278
how the molecules interact, the surface area available for interaction, and the dynamics of279
the binding process.280
• Conformational Flexibility: Peptides often have greater conformational flexibility than281
proteins. This flexibility can enable peptides to adopt multiple conformations when unbound282
and possibly select a conformation that fits a protein binding site upon interaction. Predicting283
how a peptide will conform to a protein binding site can be challenging and different from284
predicting interactions between larger, more structured proteins.285
• Interaction Surface and Specificity: Protein-protein interactions typically involve larger286
interaction surfaces and can be influenced by multiple factors, including electrostatic inter-287
actions, hydrophobic effects, and hydrogen bonding across a larger area. Peptide binding288
often involves a smaller surface area and might be driven more by specific key residues or289
motifs that fit into complementary pockets on the protein.290
• Computational Modelling: The methods for modeling and predicting these interactions291
often differ. Protein-protein interactions may require more detailed and computationally292
intensive models that account for the interactions of larger and more complex surfaces. In293
contrast, peptide binding predictions might focus more on the peptide’s ability to adopt294
conformations that match a specific binding site on a protein.295
• Biological Role and Implications: The biological implications of protein-protein versus296
protein-peptide interactions can also differ, affecting how these interactions are studied.297
Protein-protein interactions often govern major cellular processes such as signal transduction,298
metabolism, and cell structure dynamics. Protein-peptide interactions might be more299
transient and critical in regulatory mechanisms, such as those involving inhibitors, signal300
peptides, or short-lived regulatory interactions.301
• Data Availability: There might also be differences in the availability of experimental data302
for training predictive models. Protein-protein interactions have been extensively studied,303
and numerous databases contain detailed interaction data. Protein-peptide interactions,304
especially those involving short or uncommon peptides, may be less well-documented,305
making it challenging to train robust predictive models.306
Due to these differences, the strategies and algorithms developed for one type of interaction often307
need substantial modification to be effective for the other. This distinction necessitates treating these308
predictions as distinct problems, each requiring tailored approaches to model and predict binding309
affinities accurately. As such, TDC-2 defines a Protein-Peptide binding affinity prediction task310
and provides corresponding datasets and benchmarks. The formal non-generative predictive task311
9
definition is a function accepting a protein and a peptide as input and outputting a probability for312
binding affinity</p>

<p class="is-size-5">TDC-2 provides unique dataset for benchmarking this task. It contains a small training set which can be used for fine-tuning.
The majority of the data is used as a test set. As this dataset contains newly discovery ligand peptides, it is expected there will be no standard
datasets containing these samples. As such, we welcome submissions with models trained using a broad array of methodologies and datasets. More details on the dataset:

Affinity selection-mass spectrometry data of discovered ligands against single biomolecular targets (MDM2, ACE2, 12ca5) from the Pentelute Lab of MIT This dataset contains affinity selection-mass spectrometry data of discovered ligands against single biomolecular targets. Several of these AS-MS discovered ligands were taken forward for experimental validation to determine the binding affinity (KD) as measured by biolayer interferometry (BLI) to the listed target protein. If listed as a "putative binder," AS-MS alone was used to isolate the ligands to the target, with KD &lt; 1 uM required and often observed in orthogonal assays, though there is some (&lt; 50%) chance that the ligand is nonspecific. Most of the ligands are putative binders with 4446 total provided. For those characterized by BLI (only 34 total), the average KD is 266 ± 44 nM, median KD is 9.4 nM. 

Related publication: Ye X, Lee YC, Gates ZP, Ling Y, Mortensen JC, Yang FS, Lin YS, Pentelute BL. Binary combinatorial scanning reveals potent poly-alanine-substituted inhibitors of protein-protein interactions. Commun Chem. 2022 Oct 14;5(1):128. doi: 10.1038/s42004-022-00737-w. PMID: 36697672; PMCID: PMC9814900.  </p>

<p class="is-size-5">To access a benchmark in the group, use the following code:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">tdc.benchmark_group.protein_peptide_group</span> <span class="kn">import</span> <span class="n">ProteinPeptideGroup</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">ProteinPeptideGroup</span><span class="p">()</span>
<span class="n">train</span><span class="p">,</span> <span class="n">val</span> <span class="o">=</span> <span class="n">group</span><span class="p">.</span><span class="n">get_train_valid_split</span><span class="p">()</span> <span class="c1"># val dataset will be empty. use the train dataset if fine-tuning desired.
</span><span class="n">test</span> <span class="o">=</span> <span class="n">group</span><span class="p">.</span><span class="n">get_test</span><span class="p">()</span>

<span class="c1">## --- (optional) train your model --- ##
</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">model</span><span class="p">.</span><span class="n">predict</span><span class="p">(</span><span class="n">test</span><span class="p">)</span>  <span class="c1"># modify as per your model code and test output
</span><span class="n">out</span> <span class="o">=</span> <span class="n">group</span><span class="p">.</span><span class="n">evaluate</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span>
</code></pre></div></div>

<p class="is-size-5"> Follow the <b><a href="/benchmark/overview/">instructions</a></b> on how to use the <code>BenchmarkGroup</code> class and obtain training, validation, and test sets, and how to submit your model to the leaderboard. </p>

<p class="is-size-5"> The evaluation metric is AUC. More details to-be-announced.</p>

<div class="column is-12">
    <hr />
</div>

</div>
                </div>
                
            </div>
        </div>
    </section>
    
        <footer class="footer">
    <div class="container">
        
        <div class="columns is-mobile">
            <div class="column is-8 has-text-left is-vcentered">
                <a class="navbar-brand" href="/">
                    <span><img src="/tdc_horizontal.png" alt="Logo" style="max-height: 40px; max-width: 250px;"></span>
                </a>
            </div>
            <div class="column is-4 has-text-right is-vcentered">
                <a href="https://arxiv.org/abs/2102.09548">
                    <span class="icon is-large">
                      <i class="fas fa-file-alt fa-3x"></i>
                    </span>
                </a>

                <a href="https://github.com/mims-harvard/TDC">
                    <span class="icon is-large">
                      <i class="fas fab fa-github fa-3x"></i>
                    </span>
                </a>

                <a href="https://twitter.com/ProjectTDC">
                    <span class="icon is-large">
                      <i class="fas fab fa-twitter fa-3x"></i>
                    </span>
                </a>

                <a href="https://join.slack.com/t/pytdc/shared_invite/zt-x0ujg5v6-zwtQZt83fhRdgrYjXRFz5g">
                    <span class="icon is-large">
                      <i class="fas fab fa-slack fa-3x"></i>
                    </span>
                </a>
            </div>
        </div>
        
    </div>
</footer>
    
    <script src="/assets/js/app.js" type="text/javascript"></script><!-- footer scripts --></body>
</html>

